### PR TITLE
Rename "Epoch" to "Session"

### PR DESCRIPTION
**Pull request by davxy**
Superseeds https://github.com/paritytech/substrate/pull/11330 

This PR refines the work done by @gitwithterrence by addressing the issues left open.

The work mostly consists of a trivial 1 to 1 renaming from Epoch to Session of variables and structures.

In order to simplify the review:
- Exceptions to the trivial renaming where highlighted by my review comments below.
- I've not renamed YET the `client/consensus/epochs` folder.  I'm going to rename it once I have the green light.

### CODE REVIEW

File: client/consensus/babe/rpc/src/lib.rs [hunk 1]
```diff
@@ -37,23 +37,24 @@ use sp_consensus_babe::{digests::PreDigest, AuthorityId, BabeApi as BabeRuntimeA
...
 
 /// Provides rpc methods for interacting with Babe.
 #[rpc(client, server)]
 pub trait BabeApi {
-	/// Returns data about which slots (primary or secondary) can be claimed in the current epoch
+	/// Returns data about which slots (primary or secondary) can be claimed in the current session
 	/// with the keys in the keystore.
-	#[method(name = "babe_epochAuthorship")]
-	async fn epoch_authorship(&self) -> RpcResult<HashMap<AuthorityId, EpochAuthorship>>;
+	#[method(name = "babe_sessionAuthorship", aliases = ["babe_epochAuthorship"])]
```
Comment (davxy) [inline client/consensus/babe/rpc/src/lib.rs:48]: Backward compatibility by introducing previous name as a method alias

File: client/consensus/babe/src/aux_schema.rs [hunk 2]
```diff
@@ -16,24 +16,24 @@
...
 };
 use sp_blockchain::{Error as ClientError, Result as ClientResult};
 use sp_consensus_babe::{BabeBlockWeight, BabeGenesisConfiguration};
 use sp_runtime::traits::Block as BlockT;
 
-const BABE_EPOCH_CHANGES_VERSION: &[u8] = b"babe_epoch_changes_version";
-const BABE_EPOCH_CHANGES_KEY: &[u8] = b"babe_epoch_changes";
-const BABE_EPOCH_CHANGES_CURRENT_VERSION: u32 = 3;
+const SESSION_CHANGES_VERSION: &[u8] = b"babe_epoch_changes_version";
+const SESSION_CHANGES_KEY: &[u8] = b"babe_epoch_changes";
```
Comment (davxy) [inline client/consensus/babe/src/aux_schema.rs:35]: Old storage key strings should be preserved

File: frame/babe/src/lib.rs [hunk 3]
```diff
@@ -190,12 +190,12 @@ pub mod pallet {
 		InvalidConfiguration,
 	}
 
-	/// Current epoch index.
+	/// Current session index.
 	#[pallet::storage]
-	#[pallet::getter(fn epoch_index)]
+	#[pallet::getter(fn session_index)]
```
Comment (davxy) [inline frame/babe/src/lib.rs:195]: For now, for the storage, we only renamed the `getter` methods.
**NO RENAME** for the storage identifiers has been performed in this PR.

Renaming storage identifier changes the storage prefix, thus it requires a migration.

Not renamed storage identifiers: `EpochIndex`, `PendingEpochChangeConfig`, `EpochStart`, `EpochConfig`, `NextEpochConfig`.

Identifiers renaming with a migration will be performed in a follow-up PR to better isolate the change.

M

File: primitives/blockchain/src/backend.rs [hunk 4]
```diff
@@ -279,8 +279,8 @@ pub mod well_known_cache_keys {
 	/// A list of authorities.
 	pub const AUTHORITIES: Id = *b"auth";
 
-	/// Current Epoch data.
-	pub const EPOCH: Id = *b"epch";
+	/// Current Session data.
+	pub const SESSION: Id = *b"epch";
```
Comment (davxy) [inline primitives/blockchain/src/backend.rs:283]: Old key has been preserved

File: primitives/consensus/babe/src/lib.rs [hunk 5]
```diff
@@ -364,20 +368,22 @@ sp_api::decl_runtime_apis! {
...
+		/// Returns information regarding the current session.
+		#[renamed("current_epoch", 3)]
+		fn current_session() -> Session;
 
-		/// Returns information regarding the next epoch (which was already
-		/// previously announced).
-		fn next_epoch() -> Epoch;
+		/// Returns information regarding the next session (which was already announced).
+		#[renamed("next_epoch", 3)]
+		fn next_session() -> Session;
```
Comment (davxy) [inline primitives/consensus/babe/src/lib.rs:381]: API version incremented to 3.

Added a `renamed` attribute with renaming version  for renamed methods

**Review by davxy — COMMENTED** [review]
Added comments to the code that required some attention

**Comment by davxy** [comment]
If we really want to do the renaming then this is a good opportunity. It is a bit annoying renaming variables and stuff ... :-)

File: primitives/blockchain/src/backend.rs (hunk 4 above)
Comment (pepyakin) [reply primitives/blockchain/src/backend.rs:283]: Nit: maybe a comment about why it's `epch`?

**Review by sam0x17 — DISMISSED** [review]

**Comment by sam0x17** [comment]
Once the nits and merge conflicts are resolved looks good at least to me :)

**Comment by joepetrowski** [comment]
This is a confusing change IMO, as @rphmeier pointed out in the last PR, they are conceptually different. Naming them the same thing makes explaining BABE and GRANDPA more difficult. We have other naming collisions in Substrate too that have caused issues with users, e.g.:

- `Era` is in both Staking as well as the extrinsic mortality. We explain this all the time to custodians.
- `transaction_version`/`extrinsic_version` used for the serialization format and the chain's call interface (e.g. [here](https://github.com/paritytech/txwrapper-core/issues/85) and [here](https://github.com/w3f/polkadot-wiki/pull/3155#discussion_r840552316)).

**Comment by andresilva** [comment]
I'm a bit torn on this, I also agree that epochs and sessions are slightly different concepts (even though we use epochs to integrate with sessions) but don't really have any strong opinion either way. But I'm also not the target audience for this rename as I'm not confused by having these two different names. There's comments arguing for either side though.

**Comment by bkchr** [comment]
> Naming them the same thing makes explaining BABE and GRANDPA more difficult.

Can you explain this?

> pointed out in the last PR, they are conceptually different

Can someone explain this as well? What is the difference?

**Comment by andresilva** [comment]
An epoch is always measured in slots, a session is just some abstract period, e.g. aura sessions are measured in number of blocks. Session index is always contiguous while the same isn't necessarily true for epochs (since they are bound to a notion of real-world time). The concepts are tightly coupled because we do implement a session using an epoch, so there's a 1:1 mapping.

**Comment by rphmeier** [comment]
Yeah, I'm still against this change. I think it's a bad idea. 

Quoting from before
> > IMO it was a huge mistake to ever have introduced these two different namings into the code base.

> There's no guarantee that these things will be in the same code base in the future. Separating concepts decouples code, combining them (even when they aren't congruent) makes it harder to decouple in the future. You could imagine splitting the BABE logic into its own repo and then it would make more sense; it's always better to write code in a large repo in a more decoupled way.

> "This crate implements BABE consensus on top of Substrate's session abstraction"

I'm quite glad that we _did_ have the foresight to introduce these two distinct _terms_ into the codebase.

**Comment by davxy** [comment]
> aura sessions are measured in number of blocks
This is interesting.

As said somewhere else, I don't have strong opinions. But I just want to be **coherent** wrt Sassafras implementation.

In short, to avoid confusion, I would like to stick with the same terms used by Babe (since the two protocols are very closely related).
Thus:
-> If we are renaming Epoch to Session in Babe then I will be happy to name this "entity" Session in Sassafras as well.
-> If we are not renaming then I prefer to call this "Epoch" in Sassafras

It will be even more awkward to have different names in two protocols that are so strictly connected.
Not to mention that both are using `sp-epochs` crate.

So the choice here I think should be binary. All or nothing.

**Comment by rphmeier** [comment]
I agree, it should be consistent in the consensus implementations. I think it makes most sense to stick to the terminology of the paper and document that we use "Substrate Sessions to implement (Babe|Sassafras) Epochs".

Looking forward, it would be nice to split out the code of BABE and SASSAFRAS from Substrate as much as possible, for example as we've done with the finality-grandpa crate.

**Comment by rphmeier** [comment]
For example, we use the notion "voter_set" and "set_id" in GRANDPA, which is closer to the paper. Should we also rename this to session? Even those this also refers to a different mechanism from BABE Epochs. (Epochs drive sessions which drive voter sets on GRANDPA)

**Comment by bkchr** [comment]
> Looking forward, it would be nice to split out the code of BABE and SASSAFRAS from Substrate as much as possible, for example as we've done with the finality-grandpa crate.

Why? For Grandpa it makes sense as the protocol itself is much more complicated than BABE/SASSAFRAS. I mean yes we could also make this generic and move it out, but I don't see any benefit in doing this. This would only result in more work.

I would still called epochs session, as this is the BABE/SASSAFRAS implementation in the context of Substrate and as said here:

> The concepts are tightly coupled because we do implement a session using an epoch, so there's a 1:1 mapping.

When you are implementing a certain research paper, you also don't start calling all your variables using the wide variety of the Greek alphabet. The most important thing for me is consistency and consistency in Substrate would be to call it Session, not Epoch. You need to think about people approaching Substrate for the first time. These people will not start reading the BABE paper before using it. So, they will dive into the code and will see different naming for the same concept. The only thing this is doing is confusing people, nothing more. It was even said by @andresilva that we have this 1:1 mapping.

However, I will not die on this hill and I'm also okay with closing this pr.

**Comment by bkchr** [comment]
Okay, let's close this.

**Pull request closed without merging**